### ISSUE TITLE
Add "reserved" virtual functions

### ISSUE THREAD

**Issue by vadz** [wxWidgets/wxWidgets#22555]
To allow making changes to various classes without breaking ABI in 3.2.x series, we need to add stubs for "reserved" virtual functions that wouldn't do anything right now but could be repurposed for something else later -- without affecting the class ABI.

We need to do this at least for:

- [x] `wxApp`
- [x] `wxEvtHandler`
- [x] `wxWindow`
- [x] `wxTopLevelWindow`
- [x] `wxFileDialog`

Anything else?

**Comment by vadz** [comment]
It's a reference to the well-known (I think?) technique for having some virtual functions that can be reused for something useful later without breaking the ABI. As adding a virtual function is an ABI-breaking change, this can't be done in 3.2, but renaming a virtual function or adding parameters to it is _not_ an ABI change, as it doesn't change the vtbl size or layout and there should be no uses of the existing reserved function.

**Comment by paulcor** [comment]
I believe it is generally OK to add virtual functions, that is why we have `wxABI_VERSION`. Additionally, a "reserved" function is unlikely to be useful, as its signature is part of the ABI. So you can't change the name, parameters or return type.

**Comment by paulcor** [comment]
>> it is referenced by the vtable

> Not by name. Vtbl entry just contains a pointer

I've done some experiments, and I'm pretty sure it is by name, resolved to a pointer by the dynamic linker at program load time.

> Yes, I know about this document, thanks.

So then you know it says you cannot change the signature, including number of parameters, their types or qualifiers, or return type. 

Look, you might be able to get away with it using GCC (and just say there is no ABI compatibility for MSVS I guess?), but it just seems sketchy to me. I have a hard time believing anybody does this.

**Comment by vadz** [comment]
Could you please show, or at least explain what did you do in your experiments? For me, if I create a class in a shared library with 2 virtual functions, one of which is "reserved", build an application using the other (normal) virtual function, then recompile the library after changing the reserved virtual function name, type and parameters, the existing application still works perfectly fine -- which seems to confirm that it could work for us.

Note that this is not going to work in the other direction, i.e. if the library calls a virtual function defined in the executable. But I don't think there is anything we can possibly do about this.

> Look, you might be able to get away with it using GCC (and just say there is no ABI compatibility for MSVS I guess?), but it just seems sketchy to me.

Of course it's sketchy, all this is very clearly undefined behaviour. But if it helps us to be able to backport useful features to 3.2 later on, I think it would be well worth the sketchiness.

> I have a hard time believing anybody does this.

One of the very first results of a web search for "reserved virtual functions compatibility" is this [Haiku OS page](https://www.haiku-os.org/documents/dev/binary_compatibility_3_easy_steps/) which shows that this was done in BeOS ~30 years ago. It's really not a new idea.

**Comment by paulcor** [comment]
Your idea will work - with GCC - as long as the vtable is only defined in the library (meaning it has at least one non-inline virtual method). But, it might not work if the existing application has derived a class from yours. Example:

lib.h:
```c++
struct C {
    virtual int f();
#ifdef NEW
    virtual int g(int);
#else
    virtual void reserved();
#endif
};
```

lib.cpp:
```c++
#include "lib.h"

#ifdef NEW
int C::f() { return g(0); }  // existing method calls new one
int C::g(int) { return 0; }
#else
int C::f() { return 0; }
void C::reserved() { }
#endif
```

app.cpp:
```c++
#include "lib.h"

struct D: public C { };

int main()
{
    D d;
    return d.f();
}
```
```
$ g++ -Os -fpic -shared -o lib.so lib.cpp
$ g++ -Os app.cpp lib.so
$ LD_LIBRARY_PATH=$PWD ./a.out
$ g++ -Os -fpic -shared -o lib.so -DNEW lib.cpp
$ LD_LIBRARY_PATH=$PWD ./a.out
./a.out: symbol lookup error: ./a.out: undefined symbol: _ZN1C8reservedEv
$ echo _ZN1C8reservedEv | c++filt
C::reserved()
```

**Comment by vadz** [comment]
You're right, thanks, if an app-defined class derives from a library-defined class it indeed references its existing virtual methods in its vtbl. But this means we can't do it for any of the "interesting" classes, basically... Unless we go BeOS route and define our own `extern "C" ZN1C8reservedEv() { ... }`, but this is too crazy even for me.

I guess we'll have to live without the possibility of adding any virtual functions in 3.2, but I'm pretty sure it's going to be a serious limitation and I really hoped that these "reserved" functions could help us to deal with it, but we'll have to live with [`wxShadowXXX` stuff](https://github.com/wxWidgets/wxWidgets/blob/master/docs/contributing/binary-compatibility.md#workarounds-for-adding-virtual-functions) then.

**Comment by vadz** [comment]
I keep returning to this because I just can't get rid of the idea that we should be able to do something better than what we did for 3.0 -- and which clearly wasn't ideal because we hadn't been able to backport many things to it.

So what about still adding a bunch of
```cpp
virtual void WXReserved$N(void*); // N = 1..5?
```
to some key classes (those above and maybe more) to at least be able to use these functions if we need them internally, e.g. for stuff like `wxWindow::WXSetInitialFittingClientSize()`? Of course this won't be useful for the application code, but it could be useful for the implementation.

And it doesn't cost us much to do it, does it? Except for adding a few (well, 40, for 5 pointers) bytes to all virtual tables, which does add up, I guess, when multiplied by hundreds of classes, but still isn't noticeable, probably.

What do you think?

**Comment by SpareSimian** [comment]
What happened in 3.0?

I'm thinking you could make the placeholders slightly more generic by passing a parameter object (also with a generic reserved name) and then later adding a non-virtual protected macro or template wrapper to invoke them.

**Comment by vadz** [comment]
> What happened in 3.0?

Well, nothing, basically.
 
> I'm thinking you could make the placeholders slightly more generic by passing a parameter object (also with a generic reserved name) and then later adding a non-virtual protected macro or template wrapper to invoke them.

Yes, the idea would be to add a function wrapping them, i.e. if we needed to repurpose one of the functions for using an accent colour for the window title bar (just a stupid example), we'd do something like
```cpp
class wxWindow
{
public:
    ...
    bool SetTitleBarAccentColour(const wxColour& col)
    {
        return WXReserved1(&col) != NULL;
    }
};
```
and use that function in the code.

**Comment by vadz** [comment]
My idea was that it's always possible to pass a struct by pointer, i.e., continuing my stupid example
```cpp
struct SetTitleBarAccentParams {
    wxColour col;
    wxWindow* otherWindow;
    bool useGradient;
} params = { ... };
WXReserved(&params);
```
I don't think adding more parameters is really going to make it more ergonomic, will it?

**Comment by paulcor** [comment]
Well, you would have to declare that struct somewhere that both caller and implementation can see. Seems like more work to me.

**Comment by vadz** [comment]
We're going to need to define a wrapper function anyhow when we start using a previously reserved function, so declaring a struct won't add that much of overhead, I think.

BTW, I've just realized that we should make all these reserved functions protected as we'd add these wrapper functions for calling them to the class itself and only they would be public, but the functions themselves shouldn't be.

**Comment by SpareSimian** [comment]
The return value can be part of the struct. The wrapper function can extract it if needed. 
```
protected:
   struct WXReservedParams1 { ... };
   void WXReserved1(WXReservedParams1* params);
public:
   bool SetTitleBarAccentColour(const wxColour& col)
   {
      WXReservedParams1 params;
      params.col = col;
      WXReserved1(&params);
      return params.result;
   }
```

**Comment by ojwb** [comment]
With my "packaging wxWidgets for Debian" hat on:

It seems this idea must be aimed at helping people like me who really need an ABI-compatible release series to follow (especially since each development snapshot announcement says "you’re encouraged to use this release, including in production").

I'm rather luke warm on hacks involving undefined behaviour as I worry they might bite us when a new compiler version implements some new optimisation which takes advantage of that undefined behaviour.

It seems to me a better way to address the issue of not being able to add a new virtual method except in a new stable release series would simply be to start each new release series in a timely manner.  That would work better for me in so many ways.

I dug out a little table I made back in 2019 - https://groups.google.com/g/wx-dev/c/OAeDwBNPiTo/m/NJbmz-vaEAAJ:

> Looking at the time between wxWidgets .0 releases (data from
> https://en.wikipedia.org/wiki/WxWidgets since this doesn't seem to
> be easy to discover from git or the website):
>
> 2.2.0 2000-07-09
> 2.4.0 2003-01-06 ~2.5 years
> 2.6.0 2005-04-21 ~2.3 years
> 2.8.0 2006-12-13 ~1.6 years
> 3.0.0 2013-11-11 ~6.9 years
> 3.2.0 ? 5.8 years and counting 

Updating (assuming the release happens to plan):

3.2.0 2022-07-07 ~8.7 years

The Debian release schedule is not nailed to precise dates, but has a pretty firmly established 2 year cadence and for Debian packaging a new stable wx release series every 2 years or every 4 years would work well.

Ubuntu has a more strictly defined 2 year LTS release cadence so I'd guess that would work well for them too (though I'm not involved so can't really speak for them).

I'm not very familiar with other distros, but e.g. the Fedora wx maintainer also requested a more regular and more frequent wx release schedule, e.g. https://groups.google.com/g/wx-dev/c/5yzl-2KS-RY/m/Se4eb5iPBAAJ .  It seems Fedora aim to release every 6 months.

**Comment by vadz** [comment]
We'd obviously like to release more often, but this patch shouldn't prevent us from doing it (other things might, though...). And even if we do manage to release a new stable version every 2 years, it can still be useful to be able to backport a change requiring the use of a new virtual function to be implemented during this time frame.

BTW, I'm closing this because it was merged in 3.2 branch already -- but any comments are still welcome, of course.
